content/blog/2025-05-13-flat-bril/index.md

```diff
@@ -0,0 +1,170 @@
+    pub instr_labels: Option<(u32, u32)>, // Indirect: indexes into labels_idxes_store
+    pub funcs: Option<(u32, u32)>,
+}
+```
+
+```Rust 
+pub struct Function {
+    pub func_name: Vec<u8>,
+    pub func_args: Vec<FuncArg>,
+    pub func_ret_ty: Option<Type>,
+    pub var_store: Vec<u8>,
+    pub args_idxes_store: Vec<(u32, u32)>, // Intermediate array: indexes into var_store
+    pub labels_idxes_store: Vec<(u32, u32)>, // Intermediate array: indexes into labels_store
+    pub labels_store: Vec<u8>,
+    pub funcs_store: Vec<u8>,
+    pub instrs: Vec<Instr>,
+}
+```
+
+Although our function representation does contain Vec s to enable construction, these will be flattened to slices once the entire Function has been created
+
+### A flat file format
+
+Once we have flattened all of the functions and instructions in a Bril program, we are left with an array of functions, which is a full Bril program! We needed to come up with a way of writing this array to a file so that our interpreter could read the file and recover the flat data structures. To do so we first associated a table of contents with each function. As our flat function representation consists solely of arrays of bytes or arrays of indices into other arrays, what we need to record is the size of each array. We prepend this to every function, then as the table of contents has a fixed length it can be used to recover each of the function’s individual arrays. We confront a similar problem for the program itself, as it consists of an array of functions. We address this using the same approach; the beginning of every flattened bril file contains a fixed size header that stores the size of each table of contents + flat function element in the byte array that makes up the rest of the file.
+
+## Implementation
+
+### Zerocopy
+
+In order to facilitate conversion between slices of bytes and our flat data structures we used the zerocopy crate. In practice this meant adding the IntoBytes and FromBytes traits to our flat representations, and specifying their byte representations using the repr attribute. It ended up being quite challenging to get this working. Zerocopy was quite finicky about what was allowable in a struct using the zerocopy traits. We ended up needing to create new “extra-flat” versions of many of our data structures to get this to work and experimenting with different repr options. For instance we discovered that zerocopy could not convert pairs of `u32`s, or `Option`s to bytes, so we needed to create a new `I32Pair` struct that itself implemented the zerocopy traits (`I32` as opposed to `u32` because we used -1 to represent the case where the `Option` is None). We probably could have used these extra-flat data structures everywhere, rather than converting between multiple versions, but since we had written most of our flattening logic using the original data structures we decided to stick with using new extra-flat versions. 
```

Comment:
Please use a hyperlink to the zerocopy crate. And probably also to the Rust documentation for `repr`.